### ISSUE TITLE
Add indication in symbols or FakeTensor whether ranges is set via mark_dynamic

### ISSUE THREAD

**Issue by srivastavaroopak** [pytorch/pytorch#137489]
### 🚀 The feature, motivation and pitch

Currently when a tensor is marked dynamic via mark_dynamic API.  The range is reflected in the graph via ShapeEnv. However, no indication is present whether it is a user-provided range or a default range. The implication of this being hardship to differentiate user provided ranges to default ranges [2, INT_MAX]. Different backends have different implementation of dynamic ranges and it would be beneficial if a distinction is available between user set ranges and default ranges. For instance,  a backend may choose to use the user ranges and discard or manipulate default ranges.

The tensor marked dynamic has an attribute set "_dynamo_dynamic_range" which can be checked for confirming if tensor range is set by user through mark_dynamic API. However this information is available during launch(__call__) and not __init__(although graph module has ranges in shapeEnv).  The attribute however has no relation to symbols being used for that tensor and cannot be easily correlated to other input using same symbol.

The ask here is to set an indication in Symbol whether the range is set externally or internally.

### Alternatives

If information can be queried on a symbol that its range is set by the mark_dynamic API or is default, the backends can implement their logic effeciently.

cc @ezyang @chauhang @penguinwu @bobrenjc93

**Comment by srivastavaroopak** [comment]
@ezyang , Can you please share your views on the requirement.

**Comment by ezyang** [comment]
This is not hard to add but I don't really understand the use case. Also, if someone then later says `torch._check(s0 >= 4)` the range will be refined to `[4, Inf]`. Is this still a user defined range or not?

**Comment by srivastavaroopak** [comment]
The use case we're exploring involves utilizing user-provided min/max values using the mark_dynamic API to be used in the backend. We aim to focus only on user-defined ranges, as default ranges that cover a broad spectrum (e.g., [2, INF]) can lead to memory issues. As a result, we're categorizing user-defined ranges as valid and default ranges as invalid, using these distinctions to guide the creation of the valid compilation graph.

It's worth noting that the mark_dynamic API allows for both min and max values to be specified, ensuring that ranges are properly bound. If the range is later refined (e.g., to [4, Inf]) internally, rather than being provided by the user, it would not be classified as a user-defined range.

**Comment by ezyang** [comment]
In that case, why don't you just not use ranges that have an Inf max bound? If an upper bound is less than, say, sys.maxsize // 2, then just treat it as a user defined bound. There could still be false positives but there can be false positives with mark_dynamic too...

**Comment by srivastavaroopak** [comment]
You're correct and we initially performed a similar check, but over time we started noticing many false positives. One example is in the rotary_pos_embedding function, where the symbol s0 was first created with a range of [2, INT_MAX]. However, it was later updated to [2, 4096] in the default scope, which caused unexpected behavior in the backend.

Here’s what happened in more detail:

The symbol s0 was created for the size of a specific tensor dimension (L['in_input_ln213'].size()[2]), with a valid range of [2, INT_MAX].
```
create_symbol s0 = 164 for L['in_input_ln213'].size()[2] [2, 9223372036854775806] at <> in forward (_dynamo/variables/builder.py:2276 in <lambda>), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="s0"
``` 

Later, a guard was added to check if s0 <= 4096. This added a restriction to the symbol that wasn’t intended.
```
eval s0 <= 4096 [guard added] at <> in forward (_decomp/decompositions.py:763 in slice_forward), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="s0 <= 4096"
``` 

Finally, s0 was updated to have the range [2, 4096], which led to unintended behavior when processing this symbol in the backend.
```
torch/fx/experimental/symbolic_shapes.py:4697] [0/1] _update_var_to_range s0 = VR[2, 4096] (update)
``` 
This is causing the system to incorrectly assume that the tensor size min/max was set by the user, which leads to the false positives we're seeing. Hence if indication is available in symbol it will be more robust.

**Comment by ezyang** [comment]
I guess the problem is that, from my perspective, you're trying to imbue extra meaning into `mark_dynamic` API that is just... not intended. In particular, mark_dynamic with min/max isn't any different than mark_dynamic and then the user manually checking the bounds in the model, and my mental model for the API is that they should be interchangeable. And like, there's nothing wrong with me saying max=sys.maxsize//2 as a user, which will still blow out your compiler.

It really feels like you just want some dedicated API for padding hints. I'm not... fundamentally opposed to this? What makes it difficult for you to do this on your own?

**Comment by bsochack** [comment]
@ezyang, please correct me if I'm wrong. We should just propose a PR with new API to provide user hints for min/max values. Correct?

**Comment by ezyang** [comment]
Yes, but if you're proposing an entirely new API why don't you just have the user just directly tell you what they want lol

**Comment by exclamaforte** [comment]
Closing this for now, feel free to reopen with PR.
